[DEVOPS-810] Fix nix-shell env

rvl opened the pull request:
## Description

The change to `justStaticExecutablesGitRev` broke the `.env` attribute of haskell derivations.

Also there is a new nixpkgs lib function for filtering sources.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-810
https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-937

## Type of change
- [x] Build scripts/dev environment bug fix
- [x] Build scripts refactor

## QA Steps

    nix-shell default.nix -A cardano-sl-wallet-new.env
    nix-build -A all-cardano-sl

jmitchell commented on lib.nix:
```diff
@@ -12,10 +12,23 @@ let
        then result
        else default;
 
+  filterSource = src:
```
There's a `builtin.` by the same name. Using a different name, like `filterSrc`, may reduce confusion/mistakes.

rvl replied:
OK, will rename to `localLib.cleanHaskellSource`.

jmitchell commented on default.nix:
```diff
@@ -113,29 +113,7 @@ let
-            if (builtins.typeOf args.src) == "path"
-              then builtins.filterSource cleanSourceFilter args.src
-              else args.src or null;
+        src = localLib.filterSource args.src;
```
As `filterSource` is implemented I think some filters are getting dropped, like:

* `# Filter out .git repo`
* `# Filter out editor backup / swap files.`
* `# Filter out nix-build result symlinks`

Is that what we want?

rvl replied:
The good thing about `pkgs.lib.cleanSourceWith` (as opposed to `builtins.filterSource`) is that it's chainable. In `localLib.filterSource` it calls the "default" `pkgs.lib.cleanSource` function which provides the above filters. I verify this by running nix-build and checking that a cached build is used.

rvl commented on lib.nix:
```diff
@@ -12,10 +12,23 @@ let
+            # Filter out cabal build products
+            baseName == "dist" ||
+            # Filter out files which don't affect cabal build
+            lib.hasSuffix ".nix" baseName
```
Thanks for the review. The goal of these source filters is to avoid nix-build rebuilds of cabal packages when only "irrelevant" files have been changed. For example, the common case of when I edit a `.nix` file within the package's source directory. Try this:

    nix-build -A cardano-sl-wallet-new  # should download from hydra cache
    touch wallet-new/hello
    nix-build -A cardano-sl-wallet-new  # now it wants to rebuild everything

k0001 replied:
Yeah, the purpose is clear. It's just that one wouldn't normally expect a function named “clean haskell source” to pay attention to `.nix` files.

rvl replied:
OK, now it's called `localLib.cleanSourceTree`.

rvl commented:
I have added a few more useful filters to `localLib.cleanSourceTree`, and changed other instances of `builtins.filterSource` in the source code tests to use `lib.cleanSourceWith`.

disassembler commented on lib.nix:
```diff
@@ -12,10 +12,32 @@ let
+  # This is so that cached build products can be used whenever possible.
+  # It also applies the lib.cleanSource filter from nixpkgs which
+  # removes VCS directories, emacs backup files, etc.
+  cleanSourceTree = src:
```
Can we add doc files to this function as well so changing md files doesn't trigger new builds?

rvl replied:
Probably not because there is usually `extra-source-files: README.md` in the cabal file.
